File: parser/src/telemetry.ts
import * as crypto from 'node:crypto';

export interface Span {
  id: string;
  traceId: string;
  spanId: string;
  parentSpanId?: string;
  agent?: string;
  action: string;
  startTime: number;
  endTime?: number;
  durationMs?: number;
  status: 'OK' | 'ERROR' | 'UNSET';
  attributes: Record<string, any>;
}

export class TelemetryEngine {
  private activeSpans: Map<string, Span> = new Map();
  private completedSpans: Span[] = [];

  public generateTraceId(): string {
    return crypto.randomBytes(16).toString('hex');
  }

  public generateSpanId(): string {
    return crypto.randomBytes(8).toString('hex');
  }

  public startSpan(action: string, opts?: { traceId?: string; parentSpanId?: string; agent?: string; attributes?: Record<string, any> }): Span {
    const traceId = opts?.traceId || this.generateTraceId();
    const spanId = this.generateSpanId();

    const span: Span = {
      id: `span-${spanId}`,
      traceId,
      spanId,
      parentSpanId: opts?.parentSpanId,
      agent: opts?.agent,
      action,
      startTime: Date.now(),
      status: 'UNSET',
      attributes: opts?.attributes || {},
    };

    this.activeSpans.set(span.spanId, span);
    return span;
  }

  public endSpan(spanId: string, status: 'OK' | 'ERROR' = 'OK', attributes?: Record<string, any>): Span | null {
    const span = this.activeSpans.get(spanId);
    if (!span) return null;

    span.endTime = Date.now();
    span.durationMs = span.endTime - span.startTime;
    span.status = status;
    if (attributes) {
      span.attributes = { ...span.attributes, ...attributes };
    }

    this.activeSpans.delete(spanId);
    this.completedSpans.push(span);
    return span;
  }

  public injectContext(span: Span): string {
    // W3C traceparent header format: version-traceId-spanId-traceFlags
    return `00-${span.traceId}-${span.spanId}-01`;
  }

  public extractContext(traceparent: string): { traceId: string; parentSpanId: string } | null {
    const parts = traceparent.split('-');
    if (parts.length < 4 || parts[0] !== '00') return null;
    return {
      traceId: parts[1],
      parentSpanId: parts[2],
    };
  }

  public exportOTLP(): any {
    return {
      resourceSpans: [
        {
          resource: {
            attributes: [
              { key: 'service.name', value: { stringValue: 'alp-execution-engine' } },
              { key: 'telemetry.sdk.name', value: { stringValue: 'alp-telemetry' } },
            ],
          },
          scopeSpans: [
            {
              scope: { name: '@alp/telemetry', version: '17.0.0' },
              spans: this.completedSpans.map((s) => ({
                traceId: s.traceId,
                spanId: s.spanId,
                parentSpanId: s.parentSpanId || '',
                name: s.action,
                kind: 1, // SPAN_KIND_INTERNAL
                startTimeUnixNano: String(s.startTime * 1000000),
                endTimeUnixNano: String((s.endTime || s.startTime) * 1000000),
                status: { code: s.status === 'OK' ? 1 : s.status === 'ERROR' ? 2 : 0 },
                attributes: Object.entries(s.attributes).map(([k, v]) => ({
                  key: k,
                  value: { stringValue: String(v) },
                })),
              })),
            },
          ],
        },
      ],
    };
  }

  public getTraceSummary(): { totalSpans: number; activeSpans: number; okCount: number; errorCount: number } {
    const okCount = this.completedSpans.filter((s) => s.status === 'OK').length;
    const errorCount = this.completedSpans.filter((s) => s.status === 'ERROR').length;
    return {
      totalSpans: this.completedSpans.length + this.activeSpans.size,
      activeSpans: this.activeSpans.size,
      okCount,
      errorCount,
    };
  }

  public getCompletedSpans(): Span[] {
    return [...this.completedSpans];
  }
}

File: sdk/python/alp_sdk/telemetry.py
import uuid
import time
import secrets
from typing import Optional, Dict, Any, List

class Span:
    def __init__(
        self,
        action: str,
        trace_id: Optional[str] = None,
        parent_span_id: Optional[str] = None,
        agent: Optional[str] = None,
        attributes: Optional[Dict[str, Any]] = None,
    ):
        self.action = action
        self.trace_id = trace_id or secrets.token_hex(16)
        self.span_id = secrets.token_hex(8)
        self.id = f"span-{self.span_id}"
        self.parent_span_id = parent_span_id
        self.agent = agent
        self.start_time = time.time()
        self.end_time: Optional[float] = None
        self.duration_ms: Optional[float] = None
        self.status = "UNSET"
        self.attributes = attributes or {}

    def finish(self, status: str = "OK", attributes: Optional[Dict[str, Any]] = None):
        self.end_time = time.time()
        self.duration_ms = (self.end_time - self.start_time) * 1000
        self.status = status
        if attributes:
            self.attributes.update(attributes)

class TelemetryEngine:
    def __init__(self):
        self.active_spans: Dict[str, Span] = {}
        self.completed_spans: List[Span] = []

    def start_span(
        self,
        action: str,
        trace_id: Optional[str] = None,
        parent_span_id: Optional[str] = None,
        agent: Optional[str] = None,
        attributes: Optional[Dict[str, Any]] = None,
    ) -> Span:
        span = Span(
            action=action,
            trace_id=trace_id,
            parent_span_id=parent_span_id,
            agent=agent,
            attributes=attributes,
        )
        self.active_spans[span.span_id] = span
        return span

    def end_span(
        self,
        span_id: str,
        status: str = "OK",
        attributes: Optional[Dict[str, Any]] = None,
    ) -> Optional[Span]:
        span = self.active_spans.pop(span_id, None)
        if not span:
            return None
        span.finish(status=status, attributes=attributes)
        self.completed_spans.append(span)
        return span

    def inject_context(self, span: Span) -> str:
        return f"00-{span.trace_id}-{span.span_id}-01"

    def extract_context(self, traceparent: str) -> Optional[Dict[str, str]]:
        parts = traceparent.split("-")
        if len(parts) < 4 or parts[0] != "00":
            return None
        return {"trace_id": parts[1], "parent_span_id": parts[2]}

    def export_otlp(self) -> Dict[str, Any]:
        return {
            "resourceSpans": [
                {
                    "resource": {
                        "attributes": [
                            {"key": "service.name", "value": {"stringValue": "alp-python-sdk"}},
                        ]
                    },
                    "scopeSpans": [
                        {
                            "scope": {"name": "alp_sdk.telemetry", "version": "17.0.0"},
                            "spans": [
                                {
                                    "traceId": s.trace_id,
                                    "spanId": s.span_id,
                                    "parentSpanId": s.parent_span_id or "",
                                    "name": s.action,
                                    "status": {"code": 1 if s.status == "OK" else 2},
                                }
                                for s in self.completed_spans
                            ],
                        }
                    ],
                }
            ]
        }
